geth/jail/handlers.go

```diff
@@ -66,23 +66,54 @@ func registerHandlers(jail *Jail, cell common.JailCell, chatID string) error {
 	return nil
 }
 
+// makeAsyncSendHandler returns jeth.sendAsync() handler.
+func makeAsyncSendHandler(jail *Jail, cellInt common.JailCell) func(call otto.FunctionCall) otto.Value {
+	// FIXME(tiabc): Get rid of this.
+	cell := cellInt.(*Cell)
+	return func(call otto.FunctionCall) otto.Value {
+		go func() {
+			response := jail.Send(call, cell.VM)
+
+			if fn := call.Argument(1); fn.Class() == "Function" {
+				cell.Lock()
+				fn.Call(otto.NullValue(), otto.NullValue(), response)
+				cell.Unlock()
+			}
+		}()
+		return otto.UndefinedValue()
+	}
+}
+
 // makeSendHandler returns jeth.send() and jeth.sendAsync() handler
-func makeSendHandler(jail *Jail) func(call otto.FunctionCall) (response otto.Value) {
-	return jail.Send
+func makeSendHandler(jail *Jail, cellInt common.JailCell) func(call otto.FunctionCall) otto.Value {
+	// FIXME(tiabc): Get rid of this.
+	cell := cellInt.(*Cell)
+	return func(call otto.FunctionCall) otto.Value {
+		// Send calls are guaranteed to be only invoked from web3 after calling the appropriate
+		// method of jail.Cell and the cell is locked during that call. In order to allow jail.Send
+		// to perform any operations on cell.VM and not hang, we need to unlock the mutex and return
+		// it to the previous state afterwards so that the caller didn't panic doing cell.Unlock().
+		cell.Unlock()
```

Comment:
So we're bypassing the fact that cell might be blocked by other call? E.g. there is a `web3` call which hangs and `send` command is sent at the same time and this code will allow to do that?

Reply:
Yep, it's alright and there's even a test for that. The reason is that `jail.Send` now make calls through `cell.VM` - the very same VM that is locked upon the initial call. So we need to unlock it. Otherwise, it's locked 2 times during a single `jail.Call` and hangs.